### PR TITLE
Generate ext/dom class synopses based on stubs

### PR DESCRIPTION
**Pull request by kocsismate**
Generated via https://github.com/php/doc-en/pull/839

### CODE REVIEW

**Comment by kocsismate** [comment]
@salathe Can you please clarify your concerns?

**Comment by salathe** [comment]
On initially visually scanning the changes, I stopped reading when the very first file in the list (`domattr.xml`) _looked_ to be introducing mistakes. (Whether they really are mistakes, or not, remains to be seen.)

At the time, I didn't have any particular desire to continue on down that rabbit hole, so just gave a thumbs down and moved on.  I didn't reply to Máté, because I don't want to clarify my concerns.  And now, I continue to not have any particular desire to go down this rabbit hole, and so won't be giving a review _at this time_ either.

**Comment by cmb69** [comment]
A problem with reviewing this PR is that it is noisy (particularly the re-ordering of the properties). OTOH, it is done (based on?) an automated script, so later changes won't probably shuffle again. Still, a lot to review.

**Comment by kocsismate** [comment]
> A problem with reviewing this PR is that it is noisy (particularly the re-ordering of the properties). OTOH, it is done (based on?) an automated script, so later changes won't probably shuffle again. Still, a lot to review.

Yes, the command I used is ``./php-src/build/gen_stub.php --replace-classsynopses ./ext/dom ./ext/spl ./Zend/ ./ext/json  ./doc-en/reference/dom`

I agree that the property reordering makes the PR noisy. We could change the order of properties in php-src, but I'm not sure which order makes more sense, so I'm eager for suggestions. The only thing I'd like to avoid is to make manual adjustments to the generated output, since it would make subsequent class synopsis page regenerations very inconvenient.

**Comment by Girgias** [comment]
This LGTM, I think its more likely the manual was mistaken than the stubs but it might just be wise to double checked all the ``mixed`` property types, as if they are readonly I would imagine that restricting their type should also be possible.

**Review by salathe — CHANGES_REQUESTED** [review]
I managed to get past the first file this time (to 6/20 files), and there seem to be a few issues common to many files in this PR.  

See inline comments for specific details.

**General note**

Reviewing this PR _critically_ has been made more difficult than it really ought to be, due to the mixing of:
a) whitespace/formatting-only changes
b) consistency changes (e.g. adding `xi:fallback`s)
c) re-ordering of lists of things (e.g. `classref` attributes, lists of things)
d) actual content changes (e.g. new interfaces, properties, etc.)

I have very likely missed issues in the latter (let's be honest, most important) category of changes due to the others, even in the few files I've looked through thus far.   Also, if it's not clear already, please consider this review _incomplete_.  I encourage other reviewers to cast a _critial_ eye on these changes.

File: reference/dom/domattr.xml [hunk 1]
```diff
@@ -49,42 +46,47 @@
...
      <modifier>readonly</modifier>
-     <type>DOMElement</type>
-     <varname linkend="domattr.props.ownerelement">ownerElement</varname>
+     <type>bool</type>
+     <varname linkend="domattr.props.specified">specified</varname>
+     <initializer>true</initializer>
     </fieldsynopsis>
     <fieldsynopsis>
      <modifier>public</modifier>
      <modifier>readonly</modifier>
```
[the remark this answers is not in the document]
Comment (cmb69) [reply reference/dom/domattr.xml:55]: Right: https://3v4l.org/Fe0kG and https://dom.spec.whatwg.org/#interface-attr. Needs to be fixed in [the stub file](https://heap.space/xref/php-src/ext/dom/php_dom.stub.php?r=ca94d55a#288).

File: reference/dom/domcharacterdata.xml [hunk 2]
```diff
@@ -21,27 +21,29 @@
...
+     </oointerface>
     </classsynopsisinfo>
-<!-- }}} -->
- 
+
     <classsynopsisinfo role="comment">&Properties;</classsynopsisinfo>
-<!-- If the property is documented below (xml:id=domcharacterdata.props) use this -->
     <fieldsynopsis>
      <modifier>public</modifier>
+     <modifier>readonly</modifier>
```
[the remark this answers is not in the document]
Comment (cmb69) [reply reference/dom/domcharacterdata.xml:46]: Right: https://3v4l.org/UgSEZ and https://dom.spec.whatwg.org/#interface-characterdata. Needs to be fixed in the [stub file](https://heap.space/xref/php-src/ext/dom/php_dom.stub.php?r=ca94d55a#241).

File: reference/dom/domattr.xml [hunk 3]
```diff
@@ -49,42 +46,47 @@
...
+     <modifier>readonly</modifier>
+     <type>mixed</type>
+     <varname linkend="domattr.props.schematypeinfo">schemaTypeInfo</varname>
+     <initializer>null</initializer>
     </fieldsynopsis>
 
     <classsynopsisinfo role="comment">&InheritedProperties;</classsynopsisinfo>
-    <xi:include xpointer="xmlns(db=http://docbook.org/ns/docbook) xpointer(id('class.domnode')/db:partintro/db:section/db:classsynopsis/db:fieldsynopsis[preceding-sibling::db:classsynopsisinfo[1][@role='comment' and text()='&Properties;']]))">
-     <xi:fallback />
+    <xi:include xpointer="xmlns(db=http://docbook.org/ns/docbook) xpointer(id('class.domnode')/db:partintro/db:section/db:classsynopsis/db:fieldsynopsis[preceding-sibling::db:classsynopsisinfo[1][@role='comment' and text()='Properties']]))">
```
[the remark this answers is not in the document]
Comment (kocsismate) [reply reference/dom/domattr.xml:74]: No, this is not a mistake: https://github.com/php/php-src/pull/7340#discussion_r682800971 I simply didn't manage to encode the string properly, so I went without the entity. I'll try to retain `&Properties;`, and I'll appreciate any help, if anyone has a clue what to do.

File: reference/dom/domdocument.xml [hunk 4]
```diff
@@ -345,6 +364,28 @@
...
+     </listitem>
+    </varlistentry>
+    <varlistentry xml:id="domdocument.props.childelementcount">
+     <term><varname>childElementCount</varname></term>
+     <listitem>
+      <para>
+       Count of child elements.
+      </para>
+     </listitem>
+    </varlistentry>
```
[the remark this answers is not in the document]
Comment (kocsismate) [reply reference/dom/domdocument.xml:388]: Thanks for the hint, I didn't notice it's ordered alphabetically. I can update the script to generate an ordered property list, and doing so will reduce the number of changes significantly.

File: reference/dom/domattr.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (kocsismate) [reply reference/dom/domattr.xml:55]: Thanks! I've just pushed https://github.com/php/php-src/commit/cdf2f3eb165f0ceb1e58405b20146a5acdec4605 to fix both issues.

File: reference/dom/domattr.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (salathe) [reply reference/dom/domattr.xml:74]: Opened https://github.com/php/php-src/pull/7400 to address this `gen_stubs.php` issue.

File: reference/dom/domdocument.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (kocsismate) [reply reference/dom/domdocument.xml]: Thanks! I ordered the varlistentries alphabetically.

**Comment by kocsismate** [comment]
> consistency changes (e.g. adding xi:fallback)

When should I add `xi:fallbacks` elements? Currently, I always add them. What issue do extraneous `fi:fallback`s cause?

**Pull request closed without merging**